### PR TITLE
fix: CLI code generation in ng11.1+ NS-only projects

### LINKED ISSUE #311: Can't use Angular CLI with fresh nativescript installation
**Environment**
OS: Ubuntu 20.10 (Groovy Gorilla), 5.8.0-36-generic
node: 14.10.0
nativescript: 7.1.2
package.json: see repo [here](https://github.com/avif/native-test)

**Describe the bug**
I followed the guide to install the Angular CLI here [angular-cli](https://docs.nativescript.org/angular/tooling/angular-cli) with a fresh nativescript installation and could not get the Angualr CLI to work, I also could not successfully downgrade the angular version to get it to work.
On a side note, I also tried downloading the artifact from the playground but it had completely different versions and output js files laying around.

**To Reproduce**
```bash
npm i -g nativescript # nativescript@7.1.2
tns create native-test --ng
cd native-test
npm i -D @angular/cli # @angular/cli@11.1.2
npm i -D @nativescript/schematics # @nativescript/schematics@11.0.0
npx ng generate component test
An unhandled exception occurred: Cannot find module '@schematics/angular/utility/config'
Require stack:
- /home/test/Projects/native-test/node_modules/@nativescript/schematics/src/angular-project-parser.js
- /home/test/Projects/native-test/node_modules/@nativescript/schematics/src/generate/component/index.js
- /home/test/Projects/native-test/node_modules/@angular-devkit/schematics/tools/export-ref.js
- /home/test/Projects/native-test/node_modules/@angular-devkit/schematics/tools/index.js
- /home/test/Projects/native-test/node_modules/@angular/cli/utilities/json-schema.js
- /home/test/Projects/native-test/node_modules/@angular/cli/models/command-runner.js
- /home/test/Projects/native-test/node_modules/@angular/cli/lib/cli/index.js
- /home/test/Projects/native-test/node_modules/@angular/cli/lib/init.js
- /home/test/Projects/native-test/node_modules/@angular/cli/bin/ng
See "/tmp/ng-iQvXCz/angular-errors.log" for further details.
```

**Expected behavior**
A new angular component is created and linked to main module.

**Sample project**
[native-test](https://github.com/avif/native-test)

**Additional context**
I looked at `@nativescript/schematics/src/angular-project-parser.js` and it seems to fail because of changes upstream that moved `getWorkspace` from `@schematics/angular/utility/config` to `@schematics/angular/utility/workspace`, tried to change that and got this error `Cannot read property 'hello-world' of undefined` which suggests it didn't find the workspace.

### PR DESCRIPTION
**Pull request by agonper**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [x] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [x] You have signed the [CLA].
- [x] All existing tests are passing (I have not broken more 😉): https://github.com/NativeScript/nativescript-schematics/blob/master/CONTRIBUTING.md#running-tests.
- [x] Tests for the changes are included (there's no new functionality).

## What is the current behavior?
1. `@schematics/angular` v11.1+ has introduced a breaking change where `@schematics/angular/utility/config` has been removed. `@angular/cli` v11.1+ seems automatically install these new versions. In v11.1+ version of `@schematics/angular`, getWorkespace method is instead at `@schematics/angular/utility/workespace` and is async now as pointed out at #311 by @avif. Using it requires a major codebase migration. Forcing this version (~11.0.0) via regular dependencies allows to continue using `@nativescript/schematics` with NG11.1 and NG11.2 projects, but further action should be taken before NG12 arrives.
2. There was a bug in the code generation utils by which every NS NG project counted as a shared project, thus generating web components in NS-only projects. The `isWeb()` utility function was considering the existence of a `nativescript.config.ts` file as a sign of a shared project configuration, when now every NS project has this file. Instead now this decision is driven by the existence of a main.tns.ts file inside the src folder.
3. There was a regression which lead to including the `.tns` extension in all NG NS components during their generation, independently of the kind of project (shared or NS only). This extension should be used in NS component files inside shared projects only.
4. Given all NS NG templates use sass preprocessor with scss extension, it should be stated in the docs how to configure angular.json file to make this the default style sheets extension during component generation.

## What is the new behavior?
1. @schematics/angular dev dependency has been temporally promoted to become a regular dependency thus forcing a specific version range which still works with @nativescript/schematics in NG11.1+ projects as expected with no extra setup steps (like manually adding this dependency to the package.json of the project)
2. The `isWeb()`utility function has been refactored to detect the existence of a `main.tns.ts` file instead.
3. When run inside a shared project `getExtensions()` utility function returns different extensions for NS and web components, otherwise  it returns blank extension prefixes for NS-only projects.
4. Docs have been updated to reflect this possibility.

Fixes/Implements/Closes #311, #309 and #307.

BREAKING CHANGES:
None, to the best of my knowledge.

Migration steps:
None, since by merging this changes it will be possible to use @nativescript/schematics with NG11.1+ projects as expected by just following setup instructions (with no extra steps or workaround like manually installing `@schematics/angular@~11.0.0`).

----
> This is my first PR, I don't know if something else is needed (please, reach me out if so). I just wanted to help making this working back again for new projects

### CODE REVIEW

**Comment by NathanWalker** [comment]
This is amazing to see @agonper - we will bump a version update out here today with this.

**Comment by NathanWalker** [comment]
@agonper this is published with `11.2.0-rc.0` ('rc' tag) - if you could try it and if all works for you we can make final. Really appreciate your help. It was built with latest 11.2.0 and can see the additional changes that were needed here:
https://github.com/NativeScript/nativescript-schematics/commit/3ff5fa728781204999cb40a34b925a588e258343
due to getWorkspace becoming async.

**Comment by agonper** [comment]
Hi @NathanWalker, thanks for giving priority to this. I've just tested the new version and sadly it doesn't work.

I'm seeing the following error when trying to generate components and modules:
`Cannot read property 'kind' of undefined`

## Findings
I know that error is not very informative. So I've entered the rabbit hole and these are my findings:
The error is thrown by two utility functions available at `@schematics/angular/utility/ast-utils`, so this seems to be yet another breaking change in `@schematics/angular`.

In the case of the module generation it happens while calling the function `addSymbolToNgModuleMetadata` [here](https://github.com/NativeScript/nativescript-schematics/blob/3ff5fa728781204999cb40a34b925a588e258343/src/generate/module/index.ts#L310-L313).
image

In the case of the component generation it happens while calling the function `addDeclarationToModule` [here](https://github.com/NativeScript/nativescript-schematics/blob/3ff5fa728781204999cb40a34b925a588e258343/src/generate/component/ast-utils.ts#L23).
image

The later `addDeclarationToModule` ends up using the first one `addSymbolToNgModuleMetadata` ([see here](https://github.com/angular/angular-cli/blob/81678e83a7a4b525eca98d76b6678fd4b0b20068/packages/schematics/angular/utility/ast-utils.ts#L447-L452)).
image

So I've checked the `addSymbolToNgModuleMetadata` ([code](https://github.com/angular/angular-cli/blob/81678e83a7a4b525eca98d76b6678fd4b0b20068/packages/schematics/angular/utility/ast-utils.ts#L341)), to see where the problem could be. This function uses the `getDecoratorMetadata` function. Which in turn uses the `findNodes` and `_angularImportsFromNode` functions. See screenshots:
image
image

And finally the error is thrown [here](https://github.com/angular/angular-cli/blob/81678e83a7a4b525eca98d76b6678fd4b0b20068/packages/schematics/angular/utility/ast-utils.ts#L239), in the `_angularImportsFromNode` function:
image
 
It seems that the `node` is missing the `moduleSpecifier` property, hence `ms` is undefined:
```
NodeObject {
  pos: 6,
  end: 19,
  flags: 0,
  modifierFlagsCache: 0,
  transformFlags: 0,
  parent: NodeObject {
    pos: 6,
    end: 19,
    flags: 0,
    modifierFlagsCache: 0,
    transformFlags: 0,
    parent: NodeObject {
      pos: 0,
      end: 41,
      flags: 0,
      modifierFlagsCache: 0,
      transformFlags: 0,
      parent: [SourceFileObject],
      kind: 258,
      decorators: undefined,
      modifiers: undefined,
      symbol: undefined,
      localSymbol: undefined,
      locals: undefined,
      nextContainer: undefined,
      importClause: [Circular],
      moduleSpecifier: [TokenObject], <--- First appearance 
      _children: [Array]
    },
    kind: 259,
    isTypeOnly: false,
    name: undefined,
    namedBindings: [Circular],
    _children: [ [Circular] ]
  },
  kind: 261,
  elements: [
    NodeObject {
      pos: 8,
      end: 17,
      flags: 0,
      modifierFlagsCache: 0,
      transformFlags: 0,
      parent: [Circular],
      kind: 262,
      propertyName: undefined,
      name: [IdentifierObject]
    },
    pos: 8,
    end: 17,
    hasTrailingComma: false,
    transformFlags: 0
  ]
}
```
However its grand parent has it ¯\_(ツ)_/¯

## Guesses

It seems to be a TypeScript problem. I've checked in the `@schematics/angular` repo and they're using TypeScript 4.2. So I've bumped the TS version of my project to that version and it has temporally worked, but then I could not compile the project because Angular 11.2 requires a TS version lower than 4.2. However, I've tried to replicate that moment where it was temporally working and I haven't been able, there must be something else.

Hope all this information helps.

## Proposal
- Release the changes I contributed yesterday as v11.0.1 given it works with Angular 11.2 too (even when the dependencies were 11.0.0). There seems to be some interest from the community in having this working as soon as posible (3 issues, with multiple thumbs up)
- Try to work on the migration to @schematics/angular v11.2 bit by bit, since there's no real pressure, given with the previous changes it already worked with Angular 11.2 and Angular 12 has not been yet released.

**Comment by agonper** [comment]
Ok, so I've found why I couldn't reproduce the moment when it worked. It seems it was I had the schematics project linked to my project locally to be able to debug the code a bit better and that was generating conflicts with the different TypeScript versions.

If I install `@nativescript/schematics@11.2.0-rc.0` and I bump my TypeScript project version from `~4.0.0` to `~4.1.0`, it works! 🎉

The only issue I've found is that default style sheet extension configuration seems to be broken now, but I found why. I'm going to try to fix it and I'll open a PR with the changes.

So, in summary, people with NG11 projects will have to use TypeScript 4.1 instead of 4.0

**Comment by agonper** [comment]
Everything should be fixed now at #315 😃

**Comment by NathanWalker** [comment]
11.2.0 published now. Thanks again, let us know if see anything else in the final.

**Comment by agonper** [comment]
Works like a charm. Many thanks for the release @NathanWalker!
